python/note_input_widget/note_editor.py

```diff
@@ -0,0 +1,363 @@
+            # now clear the model that holds the matches
+            # this will also add a "now loading" placeholder item 
+            self._clear_model()
+    
+            # clear async download queue
+            self.__sg_data_retriever.clear()
+            
+            # kick off async data request from shotgun 
+            data = {"text": login_to_search_for}
+            self._processing_id = self.__sg_data_retriever.execute_method(self._do_sg_global_search, data)        
+            
+            # finally, show the completer and make it appear right next to 
+            # where the cursor is located.
+            cr = self.cursorRect()
+            cr.setWidth(self._completer.popup().sizeHintForColumn(0) +
+                        self._completer.popup().verticalScrollBar().sizeHint().width())
+            self._completer.complete(cr)
+
+        else:
+            # we are not in completion mode
+            # make sure completer is hidden
+            self._completer.popup().hide()
+
+
+    ##########################################################################################
+    # internal methods
+
+
+    def _clear_model(self, add_loading_item = True):
+        """
+        Clears the current data in the model.
+        
+        :param add_loading_item: if true, a "loading please wait" item will
+                                 be added.
+        """
+        # clear model
+        self._model.clear()
+        
+        if add_loading_item:
+            item = QtGui.QStandardItem("Hold on, loading...")
+            item.setData(None, self._SG_DATA_ROLE)
+            self._model.appendRow(item)
+
+    def _text_under_cursor(self):
+        """
+        Helper methods. Returns the text currently under the cursor.
+        Space and \n are considered whitespace. Example:
+        
+        Hello @john doe! 
+                ^ <---- cursor
+
+        @john doe is my friend! 
+        ^ <---- cursor
+
+        The above examples would all return '@john'
+        
+        :returns: string - see above for examples
+        """
+        cursor_pos = self.textCursor().position()
+        full_text = self.toPlainText()
+        
+        # extend selection forwards and backwards until we find whitespace
+        start_idx = cursor_pos-1
+        end_idx = cursor_pos-1
+        if end_idx < 0:
+            # special case for the first character
+            end_idx = 0
+
+        # go backwards to find start pos
+        while start_idx >= 0 and full_text[start_idx] not in (" ", "\n"):
+            start_idx = start_idx - 1
+            
+        # go forwards to find end pos. This is relevant if the cursor is 
+        # not positioned at the end of the line
+        while end_idx < len(full_text) and full_text[end_idx] not in (" ", "\n"):
+            end_idx += 1
+            
+        current_word = full_text[start_idx+1:end_idx]
+        return current_word
+
+
+    def _do_sg_global_search(self, sg, data):
```

Comment:
Ahh, this is because the `self.__sg_data_retriever.execute_method(self._do_sg_global_search, data)` call. The `execute_method()` call takes a function pointer and a dictionary as inputs. These are then passed at execution time (in a worker thread) to the method - in this case `_do_sg_global_search()`. I just used a dictonary to keep things simple - so that all options to the worker method are passed in a single data structure. That seemed easy to serialize, print out in logging etc. Hope it makes sense!

I'll add some comments to clarify!